central/securedclustercertgen/certificates.go

```diff
@@ -13,33 +13,56 @@ import (
 // secretDataMap represents data stored as part of a secret.
 type secretDataMap = map[string][]byte
 
-var v2ServiceTypes = set.NewFrozenSet[storage.ServiceType](storage.ServiceType_SCANNER_SERVICE, storage.ServiceType_SCANNER_DB_SERVICE)
-var v4ServiceTypes = set.NewFrozenSet[storage.ServiceType](storage.ServiceType_SCANNER_V4_INDEXER_SERVICE, storage.ServiceType_SCANNER_V4_DB_SERVICE)
-var allSupportedServiceTypes = func() set.FrozenSet[storage.ServiceType] {
-	types := v2ServiceTypes.Unfreeze()
-	types = types.Union(v4ServiceTypes.Unfreeze())
-	return types.Freeze()
-}()
+var scannerV2ServiceTypes = set.NewFrozenSet[storage.ServiceType](storage.ServiceType_SCANNER_SERVICE, storage.ServiceType_SCANNER_DB_SERVICE)
+var scannerV4ServiceTypes = set.NewFrozenSet[storage.ServiceType](storage.ServiceType_SCANNER_V4_INDEXER_SERVICE, storage.ServiceType_SCANNER_V4_DB_SERVICE)
+var localScannerServiceTypes = scannerV2ServiceTypes.Union(scannerV4ServiceTypes)
+
+var securedClusterServiceTypes = set.NewFrozenSet[storage.ServiceType](
+	storage.ServiceType_SENSOR_SERVICE,
+	storage.ServiceType_COLLECTOR_SERVICE,
+	storage.ServiceType_ADMISSION_CONTROL_SERVICE)
+
+var allSupportedServiceTypes = securedClusterServiceTypes.Union(localScannerServiceTypes)
+
+type certIssuerImpl struct {
+	serviceTypes set.FrozenSet[storage.ServiceType]
+}
+
+// IssueSecuredClusterCerts issues certificates for all the services of a secured cluster (including local scanner).
+func IssueSecuredClusterCerts(namespace string, clusterID string) (*storage.TypedServiceCertificateSet, error) {
+	certIssuer := certIssuerImpl{
+		serviceTypes: allSupportedServiceTypes,
+	}
+	return certIssuer.issueCertificates(namespace, clusterID)
+}
 
 // IssueLocalScannerCerts issue certificates for a local scanner running in secured clusters.
 func IssueLocalScannerCerts(namespace string, clusterID string) (*storage.TypedServiceCertificateSet, error) {
+	// In any case, generate certificates for Scanner v2.
+	serviceTypes := scannerV2ServiceTypes
+	if features.ScannerV4.Enabled() {
```

Comment:
Have you considered the following scenario:

1. Scanner V4 disabled in Central.
2. Secured Cluster is set up with V4 in AutoSense mode.
3. Certs are issued for Secured Cluster, without V4 certs.
4. V4 is enabled in Central.
5. Secured Cluster restarts/reconnects.

Would we end up in a working setup, with V4 enabled?
(Don't know from the top of my head if this would be a new issue with this change or an old one)

Reply:
These changes are old, coming from this commit: https://github.com/stackrox/stackrox/commit/3081bac1504d45e1af85ad52459e9d91aa8d0896

It's possible that it's not working correctly in all cases, but I wouldn't mix fixes to old functionality in this PR.

Reply:
BTW I think the scenario from your question should work, but would have to try it to make sure.

The localscanner cert refresh logic is always checking if the certificates are up to date. One of them missing means they are not updated and would trigger a refresh request.

So if you enable V4 in Central, it should eventually retrieve the certs on its own.

But I'd have to test to make sure that this is true.